File: Project 3/project3.py
import time
import sys

def read_file(filename,n):
    i = 0
    word_list = []
    with open(filename, "r") as f:
        for line in f:
            if i == n:
                break
            line = line.split("\n")
            word_list.append(line[0])
            i += 1
    return word_list


def first_ten(input_list):
    ten = []
    for i in input_list[0:10]:
        ten.append(i)
    return ten


def selection_sort(l):
    min_l = 0
    if len(l) < 2:
        return l
    else:
        for i in range(0, len(l)-1):
            for j in range(i+1, len(l)):
                if l[j] < l[min_l]:
                    min_l = j
            if min_l != i:
                l[i], l[min_l] = l[min_l], l[i]
    return l


def merge_sort(l):
    if len(l) < 2:
        return l
    else:
        half = int(len(l)/2)
        left = l[:half]
        right = l[half:]
        solution_left = merge_sort(left)
        solution_right = merge_sort(right)
        solution_l = merge(solution_left, solution_right)
    return solution_l


def merge(left, right):
    merged = []
    l = 0
    r = 0
    while l < len(left) and r < len(right):
        if left[l] < right[r]:
            x = left[l]
            l += 1
        else:
            x = right[r]
            r += 1
        merged.append(x)
    return merged + left[l:] + right[r:]


def main():
    if len(sys.argv) != 4:
        print('error: you must supply exactly two arguments\n\n' +
              'usage: python3 <Python source code file> <text file> <n> <selection||merge>')
        sys.exit(1)

    filename = sys.argv[1]
    n = int(sys.argv[2])
    sort_type = sys.argv[3]

    input_list = read_file(filename, n)

    print('Requested n = ' + str(n))
    print('Loaded ' + str(n) + ' lines from "' + filename + '"')
    print('First 10 words: ' + str(first_ten(input_list)))

    if sort_type == "selection":
        print('Selection Sort [O(n^2)]...')
        start = time.perf_counter()
        sorted_list = selection_sort(input_list)
        end = time.perf_counter()
    elif sort_type == "merge":
        print('Merge Sort [O(nlogn)]...')
        start = time.perf_counter()
        sorted_list = merge_sort(input_list)
        end = time.perf_counter()

    print('First 10 words: ' + str(first_ten(sorted_list)))
    print('Elapsed time: ' + str(end-start) + " seconds")

if __name__ == '__main__':
    main()

File: Project 3/project3.cpp
#include <chrono>
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

void read_file(string *list, string filename, int n){
	ifstream infile;
	infile.open(filename);
	for (int i = 0; i < n; i++){
		getline(infile, list[i]);
	}
	infile.close();
}


void first_ten(string *list, int n){
	if (n > 10)
		n = 10;
	cout << "First " << n << " words: ";
	for (int i=0; i < n; i++){
		cout << "[" << list[i] << "] ";
	}
	cout << endl;
}


void selection_sort(string *l, int n) {
    int min_l = 0;
    if (n < 2){
		return;
    }else{
		for (int i = 0; i < n - 1; ++i){
			for (int j = i + 1; j < n; ++j){
				if (l[j] < l[min_l])
					min_l = j;
			}
			if (min_l != i)
				swap(l[i], l[min_l]);
		}
    }
}


void swap(string *a, string *b){
	string swap;
	swap = *a;
	*a = *b;
	*b = swap;
}


int main(int argc, char* argv[]) {
	if (argc != 3){
		cout << "error: you must supply exactly three arguments" << endl << endl << 
			"usage: <program exe> <text file> <n>";
		return 1;
	}

	string filename = string(argv[1]);
	int n = atoi(argv[2]);
	string * input_list;
	input_list = new string [n];

	read_file(input_list, filename, n);

	cout << "Requested n = " << n << endl;
	cout << "Loaded " << n << " line from '" << filename << "'" << endl;
	first_ten(input_list, n);

    cout << "Selection Sort [O(n2)]..." << endl;
    auto start = chrono::high_resolution_clock::now();
    selection_sort(input_list, n);
    auto end = chrono::high_resolution_clock::now();
	first_ten(input_list, n);
    int microseconds = chrono::duration_cast<chrono::microseconds>(end - start).count();
    double seconds = microseconds / 1E6;
    cout << "elapsed time: " << seconds << " seconds" << endl;

    return 0;
}